### ISSUE TITLE
Get list of opened sites

### ISSUE THREAD

**Issue by sprnza** [luakit/luakit#334]
Hi there!
Is there a way to get all opened websites in all tabs? From commandline?

**Comment by aidanholm** [comment]
Hey :)
Do you mean something like a `:buffer` command, that shows you all open windows+tabs, and lets you switch between them? Nothing like that currently, but it should be pretty simple to add; I'll have a go after I finish up the new release.

**Comment by sprnza** [comment]
No, I mean something like

```
luakit --buffer | grep -i "youtube"
```

while another instance of luakit is running.
I'd like to have this feature to be able to switch off DPMS when youtube is opened on a tab wich is currently being viewed by user, for example.

**Comment by aidanholm** [comment]
Oh, I see. It'd be tricky to retrieve data from the primary instance, although you can send it commands, e.g. to tell it to dump a list of buffers to a file.

If you only care about visible tabs in all windows, and only whether a youtube URI is open, you could probably do it as a small Lua module, something like this:
```lua
local prev_yt_visible 
local rescan = function ()
    local yt_visible = false
    for _, w in pairs(window.bywidget) do
        if w.view.uri:find("youtube.com") then yt_visible = true end
    end
    if yt_visible ~= prev_yt_visible then
        luakit.spawn("DPMS command", yt_visible and "disable" or "enable")
        prev_yt_visible = yt_visible
    end
end
window.add_signal("init", function (w)
    w.tabs:add_signal("switch-page", rescan)
end)
webview.add_signal("init", function (view)
    view:add_signal("property::uri", rescan)
end)
```
This just checks the URIs of the currently visible tabs, every time you change tab or change a tab's 
URI. Haven't tested this, but it should work: you'll need to fix the command in the `luakit.spawn` line.

**Comment by sprnza** [comment]
Thank you!

**Comment by aidanholm** [comment]
You're welcome!

**Comment by sprnza** [comment]
Seems like `w.tabs:add_signal("switch-page", rescan)` is not being fired on tab switch.

**Comment by aidanholm** [comment]
That's strange. If you start luakit, run `:lua w.tabs:add_signal("switch-page", print)` and then start switching pages, is there any output? Where in your config did you add the code?

But in fact, it'd be better to not attach to that signal (or the window `init` signal), and instead attach to the view `switched-page` signal, which is fired on the webview _after_ the switch finishes:
```lua
webview.add_signal("init", function (view)
    view:add_signal("property::uri", rescan)
    view:add_signal("switched-page", rescan)
end)
```
That should work better

**Comment by sprnza** [comment]
> That's strange. If you start luakit, run :lua w.tabs:add_signal("switch-page", print) and then start switching pages, is there any output? Where in your config did you add the code?

Do you mean stdout? If so running this command gives nothing on stdout. (neither for new nor for loaded with a session (see below) tabs)

```
:lua window = require("window"); window.add_signal("init", function (w) w.tabs:add_signal("switch-page", print) end)
```
 Seems like `vew:add_signal("switched-page", callback)` get the callback run only for new opened tabs, if I switch back to a tab wich was loaded with a session it doesn't run. 
And if I understand correctly since `"switched-page"` is fired on a tab load finish there is no point in `"property:uri"`. Right?

Thank you for your help!

**Comment by aidanholm** [comment]
Yes, stdout. What version of luakit are you running? If you're using a recent version, there's an automatic variable `w` you can use in lua commands that is the current window table, so you can directly access `w.tabs` from a `:lua` command. The window `init` signal only fires when opening a _new_ window, so your command will only print when switching tabs on a window created after you run it.

`view:add_signal()` is being called from the webview init signal, so it is only called for new tabs. If you restart the browser, however, it'll be called for all of them. `switched-page` is _only_ called after changing the currently visible tab, _not_ when a tab finishes loading. So `property::uri` is still necessary.

**Comment by sprnza** [comment]
Yeah, I was using slightly outdated version. Just got updated from AUR (luakit-git) and now pages don't get loaded with this error:

```
[    0.146880] V: luaH_widget_set_type:224: created widget of type: spinner
Error loading module '/usr/share/luakit/luakit.so': /usr/share/luakit/luakit.so: undefined symbol: ipc_recv_page_created
Error loading module '/usr/share/luakit/luakit.so': /usr/share/luakit/luakit.so: undefined symbol: ipc_recv_page_created
```

```
➜  ~ pacaur -Qs luakit
local/luakit-git r3526.0c77c93c-1
```

**Comment by sprnza** [comment]
variable `w` is still a nil value

**Comment by aidanholm** [comment]
Issue fixed in IRC: load error fixed at cbcd08c5
